src/hal/hal.cpp

```diff
@@ -384,16 +384,27 @@ bool hal_init_with_pinmap(const HalPinmap_t *pPinmap)
     }
 }; // namespace Arduino_LMIC
 
+
+static hal_failure_handler_t* custom_hal_failure_handler = NULL;
+
 void hal_failed (const char *file, u2_t line) {
+    if (custom_hal_failure_handler != NULL) {
+        (*custom_hal_failure_handler)(file, line);
+    } else {
```

Comment:
Don't we want a while-loop before the `else`? Otherwise `hal_failed()` might return, which could cause chaos. Or annotate `hal_failure_handler_t` with `__attribute__((__noreturn__))` when declaring.  Of course, we also ought to annotate `hal_failed()` the same way.  Using `__attribute__` is unfortunately a GCC-ism, and to fix this, we'd need to add yet more of the MCCI portability framework in lmic_env.h which can't happen today.

The simplest thing is to rearrange `hal_failed` slightly so that the `hal_disableIRQts(); while(1);` is outside the `if/else`; then we can deal with portability to more compilers, etc., after we get through the port 224 issues.

Reply:
You are absolutely right.